src/cache/core/types/common.ts

```diff
@@ -88,8 +93,28 @@ export type ModifierDetails = {
   storage: StorageType;
 }
 
-export type Modifier<T> = (value: T, details: ModifierDetails) => T;
-
-export type Modifiers = {
-  [fieldName: string]: Modifier<any>;
-};
+export type Modifier<T> = (
+  value: T,
+  details: ModifierDetails
+) => T | DeleteModifier | InvalidateModifier;
+
+type StoreObjectValueMaybeReference<StoreVal> = 
+  StoreVal extends Record<string, any>[]
+  ? Readonly<StoreVal> | readonly Reference[]
+  : StoreVal extends Record<string, any>
+  ? StoreVal | Reference
+  : StoreVal;
+
+export type AllFieldsModifier<
+  Entity extends Record<string, any>
+> = Modifier<Entity[keyof Entity] extends infer Value ? 
+  StoreObjectValueMaybeReference<Exclude<Value, undefined>>
+  : never>;
```

Comment:
Is there a benefit to using a conditional type with `infer` over just using `Entity[keyof Entity]` directly?

```suggestion
> = Modifier<
  StoreObjectValueMaybeReference<Exclude<Entity[keyof Entity], undefined>>
  >;
```

I used this code and didn't see any typechecking errors in the new tests you added

Reply:
it forces to run `StoreObjectValueMaybeReference` on each member of the union type instead of the full union type.

So, for `Entity = { a: string; b: number }`, my version will run
`StoreObjectValueMaybeReference<string> | StoreObjectValueMaybeReference<number>`
while your version runs
`StoreObjectValueMaybeReference<string | number>`
which *might* infer with the `extends` check within `StoreObjectValueMaybeReference`, so I went this way to better make 100% sure.

Reply:
Interesting. Thanks for clarifying :+1: